File: showroom/user/python/bitstest.py
#!/usr/bin/python

from rtai_lxrt import *
from rtai_scb  import *
from rtai_bits import *

from math import *
from array import *
from thread import *
from random import *

SLEEP_TIME = 200000
BITS_TIMOUT = 1

bits = c_void_p()

def fun0(null1, null2) :
	resulting_mask = c_ulong(0)

	task = rt_task_init_schmod(nam2num("TASK0"), 0, 0, 0, 0, 0xF)
	rt_make_hard_real_time()
	print "FUN0 WAITS FOR ALL_SET TO 0x0000FFFF WHILE BITS MASK IS ", hex(rt_get_bits(bits))
	rt_bits_wait(bits, ALL_SET, 0x0000FFFF, NOP_BITS, 0, byref(resulting_mask))
	print "FUN0 RESUMED WITH BITS MASK ", hex(rt_get_bits(bits))
	rt_sleep(nano2count(SLEEP_TIME))
	print "FUN0 WAITS_IF FOR ALL_SET TO 0xFFFF0000 WHILE BITS MASK IS ", hex(rt_get_bits(bits))
	rt_bits_wait_if(bits, ALL_SET, 0xFFFF0000, NOP_BITS, 0, byref(resulting_mask))
	print "FUN0 WAITS FOR ALL_SET TO 0xFFFF0000 WHILE BITS MASK IS ", hex(rt_get_bits(bits))
	rt_bits_wait(bits, ALL_SET, 0xFFFF0000, NOP_BITS, 0, byref(resulting_mask))
	print "FUN0 RESUMED WITH BITS MASK ", hex(rt_get_bits(bits))
	rt_sleep(nano2count(SLEEP_TIME))
	print "FUN0 WAITS 2 s FOR ALL_SET TO 0xFFFFFFFF WHILE BITS MASK IS ", hex(rt_get_bits(bits))
	timeout = rt_bits_wait_timed(bits, ALL_SET, 0xFFFFFFFF, NOP_BITS, 0, nano2count(2000000000), byref(resulting_mask))
	if timeout == BITS_TIMOUT :
		print "FUN0 RESUMED WITH BITS MASK ", hex(rt_get_bits(bits)), "(TIMEOUT)"
	else :
		print "FUN0 RESUMED WITH BITS MASK ", hex(rt_get_bits(bits))
	rt_make_soft_real_time()
	rt_task_delete(task)
	return

def fun1(null1, null2) :
	resulting_mask = c_ulong(0)

	task = rt_task_init_schmod(nam2num("TASK1"), 0, 0, 0, 0, 0xF)
	rt_make_hard_real_time()
	print "FUN1 WAITS FOR ALL_CLR TO 0xFFFF0000 WHILE BITS MASK IS ", hex(rt_get_bits(bits))
	rt_bits_wait(bits, ALL_CLR, 0xFFFF0000, NOP_BITS, 0, byref(resulting_mask))
	print "FUN1 RESUMED WITH BITS MASK ", hex(rt_get_bits(bits))
	rt_sleep(nano2count(SLEEP_TIME))
	print "FUN1 WAITS_IF FOR ALL_SET TO 0xFFFF0000 WHILE BITS MASK IS ", hex(rt_get_bits(bits))
	rt_bits_wait_if(bits, ALL_SET, 0xFFFF0000, NOP_BITS, 0, byref(resulting_mask))
	print "FUN1 WAITS FOR ALL_SET TO 0xFFFF0000 WHILE BITS MASK IS ", hex(rt_get_bits(bits))
	rt_bits_wait(bits, ALL_SET, 0xFFFF0000, NOP_BITS, 0, byref(resulting_mask))
	print "FUN1 RESUMED WITH BITS MASK ", hex(rt_get_bits(bits))
	rt_sleep(nano2count(SLEEP_TIME))
	print "FUN1 WAITS 2 s FOR ALL_SET TO 0xFFFFFFFF WHILE BITS MASK IS ", hex(rt_get_bits(bits))
	timeout = rt_bits_wait_timed(bits, ALL_SET, 0xFFFFFFFF, NOP_BITS, 0, nano2count(2000000000), byref(resulting_mask))
	if timeout == BITS_TIMOUT :
		print "FUN1 RESUMED WITH BITS MASK ", hex(rt_get_bits(bits)), "(TIMEOUT)"
	else :
		print "FUN1 RESUMED WITH BITS MASK ", hex(rt_get_bits(bits))
#	rt_make_soft_real_time()
	rt_task_delete(task)
	return

def fun(null1, null2) :
	task = rt_task_init_schmod(nam2num("TASK"), 0, 0, 0, 0, 0xF)
	rt_make_hard_real_time()
	mask = hex(rt_bits_signal(bits, SET_BITS, 0x0000FFFF))
	print "SIGNAL SET_BITS 0x0000FFFF AND RETURNS BITS MASK ", mask
	rt_sleep(nano2count(SLEEP_TIME))
	mask = hex(rt_bits_signal(bits, CLR_BITS, 0xFFFF0000))
	print "SIGNAL CLR_BITS 0xFFFF0000 AND RETURNS BITS MASK ", mask
	rt_sleep(nano2count(SLEEP_TIME))
	print "RESET BITS TO 0"
	rt_bits_reset(bits, 0)
	rt_sleep(nano2count(SLEEP_TIME))
	rt_bits_reset(bits, 0)
	rt_make_soft_real_time()
	rt_task_delete(task)
	return


task = rt_task_init_schmod(nam2num("MYTASK"), 9, 0, 0, 0, 0xF)
bits = rt_bits_init(nam2num("BITS"), 0xFFFF0000)
start_rt_timer(0)
start_new_thread(fun0, (NULL, NULL))
rt_sleep(nano2count(SLEEP_TIME*10))
start_new_thread(fun1, (NULL, NULL))
rt_sleep(nano2count(SLEEP_TIME*10))
start_new_thread(fun, (NULL, NULL))
#print "TYPE <ENTER> TO TERMINATE"
libc.getchar()
stop_rt_timer()
rt_bits_delete(bits)
rt_task_delete(task)

File: showroom/user/bitstest/bitstest.c
/*
COPYRIGHT (C) 2002  Paolo Mantegazza <mantegazza@aero.polimi.it>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
*/


#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <sys/mman.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/time.h>

#include <rtai_bits.h>

#define SLEEP_TIME  200000

static int thread0, thread1, thread;

static BITS *bits;

static void *fun0(void *arg)
{
	RT_TASK *task;
	unsigned long resulting_mask;
	int timeout;

	task = rt_task_init_schmod(nam2num("TASK0"), 0, 0, 0, SCHED_FIFO, 0xF);
	mlockall(MCL_CURRENT | MCL_FUTURE);
	rt_make_hard_real_time();
	rt_printk("FUN0 WAITS FOR ALL_SET TO 0x0000FFFF WHILE BITS MASK IS %lx\n", rt_get_bits(bits));
	rt_bits_wait(bits, ALL_SET, 0x0000FFFF, NOP_BITS, 0, &resulting_mask);
	rt_printk("FUN0 RESUMED WITH BITS MASK %lx\n", rt_get_bits(bits));
	rt_sleep(nano2count(SLEEP_TIME));
	rt_printk("FUN0 WAITS_IF FOR ALL_SET TO 0xFFFF0000 WHILE BITS MASK IS %lx\n", rt_get_bits(bits));
	rt_bits_wait_if(bits, ALL_SET, 0xFFFF0000, NOP_BITS, 0, &resulting_mask);
	rt_printk("FUN0 WAITS FOR ALL_SET TO 0xFFFF0000 WHILE BITS MASK IS %lx\n", rt_get_bits(bits));
	rt_bits_wait(bits, ALL_SET, 0xFFFF0000, NOP_BITS, 0, &resulting_mask);
	rt_printk("FUN0 RESUMED WITH BITS MASK %lx\n", rt_get_bits(bits));
	rt_sleep(nano2count(SLEEP_TIME));
	rt_printk("FUN0 WAITS 2 s FOR ALL_SET TO 0xFFFFFFFF WHILE BITS MASK IS %lx\n", rt_get_bits(bits));
	timeout = rt_bits_wait_timed(bits, ALL_SET, 0xFFFFFFFF, NOP_BITS, 0, nano2count(2000000000LL), &resulting_mask);
	rt_printk("FUN0 RESUMED WITH BITS MASK %lx %s\n", rt_get_bits(bits), timeout == BITS_TIMOUT ? "(TIMEOUT)" : "\0");
	rt_make_soft_real_time();
	rt_task_delete(task);
	return 0;
}

static void *fun1(void *arg)
{
	RT_TASK *task;
	unsigned long resulting_mask;
	int timeout;

	task = rt_task_init_schmod(nam2num("TASK1"), 0, 0, 0, SCHED_FIFO, 0xF);
	mlockall(MCL_CURRENT | MCL_FUTURE);
	rt_make_hard_real_time();
	rt_printk("FUN1 WAITS FOR ALL_CLR TO 0xFFFF0000 WHILE BITS MASK IS %lx\n", rt_get_bits(bits));
	rt_bits_wait(bits, ALL_CLR, 0xFFFF0000, NOP_BITS, 0, &resulting_mask);
	rt_printk("FUN1 RESUMED WITH BITS MASK %lx\n", rt_get_bits(bits));
	rt_sleep(nano2count(SLEEP_TIME));
	rt_printk("FUN1 WAITS_IF FOR ALL_SET TO 0xFFFF0000 WHILE BITS MASK IS %lx\n", rt_get_bits(bits));
	rt_bits_wait_if(bits, ALL_SET, 0xFFFF0000, NOP_BITS, 0, &resulting_mask);
	rt_printk("FUN1 WAITS FOR ALL_SET TO 0xFFFF0000 WHILE BITS MASK IS %lx\n", rt_get_bits(bits));
	rt_bits_wait(bits, ALL_SET, 0xFFFF0000, NOP_BITS, 0, &resulting_mask);
	rt_printk("FUN1 RESUMED WITH BITS MASK %lx\n", rt_get_bits(bits));
	rt_sleep(nano2count(SLEEP_TIME));
	rt_printk("FUN1 WAITS 2 s FOR ALL_SET TO 0xFFFFFFFF WHILE BITS MASK IS %lx\n", rt_get_bits(bits));
	timeout = rt_bits_wait_timed(bits, ALL_SET, 0xFFFFFFFF, NOP_BITS, 0, nano2count(2000000000LL), &resulting_mask);
	rt_printk("FUN1 RESUMED WITH BITS MASK %lx %s\n", rt_get_bits(bits), timeout == BITS_TIMOUT ? "(TIMEOUT)" : "\0");
	rt_make_soft_real_time();
	rt_task_delete(task);
	return 0;
}

static void *fun(void *arg)
{
	RT_TASK *task;

	task = rt_task_init_schmod(nam2num("TASK"), 0, 0, 0, SCHED_FIFO, 0xF);
	mlockall(MCL_CURRENT | MCL_FUTURE);
	rt_make_hard_real_time();
	rt_printk("SIGNAL SET_BITS 0x0000FFFF AND RETURNS BITS MASK %lx\n",
	rt_bits_signal(bits, SET_BITS, 0x0000FFFF));
	rt_sleep(nano2count(SLEEP_TIME));
	rt_printk("SIGNAL CLR_BITS 0xFFFF0000 AND RETURNS BITS MASK %lx\n",
	rt_bits_signal(bits, CLR_BITS, 0xFFFF0000));
	rt_sleep(nano2count(SLEEP_TIME));
	rt_printk("RESET BITS TO 0\n");
	rt_bits_reset(bits, 0);
	rt_sleep(nano2count(SLEEP_TIME));
	rt_bits_reset(bits, 0);
	rt_make_soft_real_time();
	rt_task_delete(task);
	return 0;
}

int main(void)
{
	RT_TASK *task;

	task = rt_task_init_schmod(nam2num("MYTASK"), 9, 0, 0, SCHED_FIFO, 0xF);
	bits = rt_bits_init(nam2num("BITS"), 0xFFFF0000);
	rt_set_oneshot_mode();
	start_rt_timer(nano2count(SLEEP_TIME));
	thread0 = rt_thread_create(fun0, NULL, 10000);
	rt_sleep(nano2count(SLEEP_TIME));
	thread1 = rt_thread_create(fun1, NULL, 10000);
	rt_sleep(nano2count(SLEEP_TIME));
	thread = rt_thread_create(fun, NULL, 10000);
	printf("TYPE <ENTER> TO TERMINATE\n");
	getchar();
	stop_rt_timer();
	rt_bits_delete(bits);
	rt_task_delete(task);
	return 0;
}
